### PR TITLE
feat: add pnpm workspace support

### LINKED ISSUE #43: Request: support workspaces.
Our project is configured ad a monorepo using Pnpm workspaces. 

in our projects we import local packages like so 

image

We would like to see these resolve to the version in our local package.

### PR DESCRIPTION
**Pull request by thecannabisapp**
This PR adds support for pnpm workspaces, resolving both `workspace:` and `catalog:` protocol references so that update decorations work correctly in monorepo setups.

### What it does
- **`workspace:*`, `workspace:^`, `workspace:~`, `workspace:x.y.z`** — resolves the referenced package to its actual version from the local workspace package.json, then checks npm for available updates.
- **`catalog:`, `catalog:default`, `catalog:<name>`** — resolves versions from the root `pnpm-workspace.yaml` `catalog` and `catalogs` sections, then checks npm for available updates.
- Skips quick-fix/bulk updates for workspace and catalog deps (the version lives in the workspace package or catalog definition, so updating the reference directly would be incorrect).
- Caches workspace discovery and catalog parsing for performance.

### Test coverage
- 52 unit tests covering workspace resolution, catalog resolution, glob patterns, YAML edge cases (comments, mixed quotes), and package.json parsing integration.

Closes #43
Closes #61

### CODE REVIEW

**Comment by thecannabisapp** [comment]
Hi @pgsandstrom — whenever you have a moment, this PR is ready for your review. I've added both workspace and catalog resolution with full test coverage, and the existing vscode e2e test still passes. Happy to make any changes you'd like. Thanks for maintaining the extension!

**Comment by pgsandstrom** [comment]
Thanks for the PR.

1. Could you merge in the latest changes from the master branch? This is on me, I actually forgot to push the latest version to github.
2. manual yarn parsing is fragile, it is going to fail for some users.
3. Iterate a few more times with AI. Ask it like "make a critical review of my changes". And then manually review the suggestions to see what is sensible.
4. Go through the code manually and check what comments and abstractions that are sensible. AI does not have a good sense for what needs explanation, this needs to be done by a human unless we want code smell.

**Comment by thecannabisapp** [comment]
Hi @pgsandstrom — thanks for the review! I've gone through all four points and pushed the updates. Here's what changed:

**1. Merged latest master ✅**
Pulled in your v3.5.1 release + the config-reading bugfix and test.

**2. Replaced fragile manual YAML parsing ✅**
You were absolutely right — the hand-rolled line-by-line parsers for  would break on inline arrays, flow styles, anchors, multi-line strings, etc. I replaced all of it with  (now a direct dependency). The ~200 lines of  /  /  became ~40 lines of normal object property access. Also added support for  as a fallback.

**3. Iterated with critical AI review ✅**
Did two more passes. Found a real bug during the second one:

> **Bug:**  dependencies were never fetched from npm because  only resolved  versions before the semver filter. A raw  string fails , so it was silently dropped — meaning catalog deps showed no update decorations at all.

Fixed by adding catalog resolution in  alongside workspace resolution. Also added workspace root caching to avoid repeated directory traversal, hardened tests with , and added coverage for  extension and invalid YAML.

**4. Cleaned up comments and abstractions ✅**
Removed verbose JSDoc that restated the obvious (e.g. "Returns undefined if the version is not a workspace reference or cannot be resolved" on a function literally named  returning ). Kept the meaningful ones — like the "Skip quick-fix upgrades for workspace dependencies" comments in  /  that explain *why*, not *what*.

Also removed the internal-only  /  interfaces that were artifacts of the manual parsing approach.

All 56 tests pass, lint and typecheck are clean, and the existing vscode e2e test still passes. Ready for another look whenever you have time!

**Comment by thecannabisapp** [comment]
Quick follow-up — the backticks in my last comment got eaten by shell escaping. The key changes were:

- Replaced ~200 lines of hand-rolled YAML line parsers with js-yaml (proper parser, handles all YAML edge cases)
- Fixed a real bug: catalog: deps were never fetched from npm because refreshPackageJsonData only resolved workspace: versions before the semver filter. Added catalog resolution there too.
- Added workspace root caching, pnpm-workspace.yml support, and try/finally in tests.
- Removed verbose JSDoc and internal-only BlockResult / NestedBlockResult interfaces.

Full diff: https://github.com/thecannabisapp/package-json-upgrade/compare/f8c6a01...ca1daae

File: src/updateAction.ts [hunk 1]
```diff
@@ -21,10 +21,21 @@ export class UpdateAction implements vscode.CodeActionProvider {
...
+    // Skip quick-fix upgrades for workspace dependencies — the version lives in the workspace package
+    if (dep.isWorkspace === true) {
+      return
+    }
+
+    // Skip quick-fix upgrades for catalog dependencies — the version lives in the workspace catalog
+    if (dep.isCatalog === true) {
+      return
+    }
+
```
Comment (michaelfaith) [inline src/updateAction.ts]: Nit: why not combine these three bailout conditions into a single `if`?

File: src/updateAll.ts [hunk 2]
```diff
@@ -32,6 +32,16 @@ export const updateAll = (textEditor?: vscode.TextEditor): UpdateEdit[] => {
...
+        // Skip workspace dependencies — the version lives in the workspace package
+        if (dep.isWorkspace === true) {
+          return
+        }
+
+        // Skip catalog dependencies — the version lives in the workspace catalog
+        if (dep.isCatalog === true) {
+          return
+        }
+
```
Comment (michaelfaith) [inline src/updateAll.ts]: Nit: similar feedback here.  Why not combine these two into a single condition?

File: src/workspace.ts [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+  let dir = path.dirname(packageJsonPath)
+  while (dir !== path.dirname(dir)) {
+    if (fs.existsSync(path.join(dir, 'pnpm-workspace.yaml'))) {
+      workspaceRootCache.set(packageJsonPath, dir)
+      return dir
+    }
+    if (fs.existsSync(path.join(dir, 'pnpm-workspace.yml'))) {
+      workspaceRootCache.set(packageJsonPath, dir)
+      return dir
+    }
```
Comment (michaelfaith) [inline src/workspace.ts]: Similar feedback: recommend combining these into one `if`.

File: src/workspace.ts [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    return results
+  }
+
+  const nextPath = path.join(currentPath, head)
+  if (!fs.existsSync(nextPath)) {
+    return []
+  }
+  return resolveGlobRecursive(nextPath, tail)
+}
```
Comment (michaelfaith) [inline src/workspace.ts]: I would recommend removing all of this and just using something like tinyglobby https://www.npmjs.com/package/tinyglobby.

**Review by michaelfaith — COMMENTED** [review]
@pgsandstrom this is pretty solid 👍 . I highlighted a few minor things, and then one recommendation to use a standard glob api, rather than rolling their own.  Otherwise, this seems reasonable.

**Comment by thecannabisapp** [comment]
@michaelfaith thanks for the review — great catches.

**Nits:** Agreed on all three. I'll flatten the bailouts in `updateAction.ts` / `updateAll.ts` and the nested catalog lookup in `workspace.ts` with optional chaining / nullish coalescing. Much cleaner.

**tinyglobby:** Also agreed. The hand-rolled glob only handles `*` and `**`, but pnpm workspace patterns support full micromatch semantics (brace expansion, negation, extglob, etc). I'll swap `resolveGlob` / `resolveGlobRecursive` for `tinyglobby` — keeps the bundle small and gives us correctness.

@pgsandstrom this should address the remaining feedback. Let me know if you'd like anything else tweaked.

**Comment by pgsandstrom** [comment]
@michaelfaith could you have a last look over to validate the changes? If you give it an okay, I will do a super quick look through and then I merge and release.

File: src/workspace.ts [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+  let dir = path.dirname(packageJsonPath)
+  while (dir !== path.dirname(dir)) {
+    if (fs.existsSync(path.join(dir, 'pnpm-workspace.yaml'))) {
+      workspaceRootCache.set(packageJsonPath, dir)
+      return dir
+    }
+    if (fs.existsSync(path.join(dir, 'pnpm-workspace.yml'))) {
+      workspaceRootCache.set(packageJsonPath, dir)
+      return dir
+    }
```
Comment (michaelfaith) [inline src/workspace.ts]: @thecannabisapp It doesn't look like this one was addressed.  Was it missed?

**Review by michaelfaith — APPROVED** [review]
@pgsandstrom looks good to me.  I just pointed out one of the Nits not being resolved (possibly missed a commit?).  But I don't  consider that blocking.

**Comment by thecannabisapp** [comment]
my bad, apologies. That one was indeed missed. Fixed now:

- `updateAction.ts`: combined all three bailouts (`dep === undefined`, `isWorkspace`, `isCatalog`) into a single `if`
- `workspace.ts`: deduplicated the `.yaml`/`.yml` checks in `findPnpmWorkspaceRoot` by reusing `findWorkspaceFile`, and refactored `findWorkspaceFile` itself to loop over extensions instead of two sequential `if` blocks

Pushed to master. Thanks again for the thorough review.

**Comment by pgsandstrom** [comment]
Okay, I had to dig in as to why tests just sometimes was failing randomly. Its because the tests write to the same temporary file, and they run in parallell. However, Im just gonna fix it myself in the main branch. Thank you very much for everyone, especially @michaelfaith, for helping out!
